File: leetcode/first-unique-character-in-a-string.py
# 387. First Unique Character in a String
# 🟢 Easy
#
# https://leetcode.com/problems/first-unique-character-in-a-string/
#
# Tags: Hash Table - String - Queue - Counting

import timeit
from collections import Counter


# One pass to count character frequencies and one pass to find the
# first character that is unique in the string.
#
# Time complexity: O(n) - Two passes == O(2n) => O(n).
# Space complexity: O(1) - The counter has a max size of 26.
#
# Runtime: 160 ms, faster than 59.19%
# Memory Usage: 14.1 MB, less than 59.01%
class Solution:
    def firstUniqChar(self, s: str) -> int:
        # One pass to count the character frequencies.
        freq = Counter(s)
        # Second pass to return the index of the first character with
        # frequency == 1.
        for i, c in enumerate(s):
            if freq[c] == 1:
                return i
        # If no character had a frequency of 1, return -1
        return -1


# This solution runs faster because the work is being done by C code
# instead of a Python loop. If the code performing the work was both C
# or Python, it should be slower than the previous one.
#
# Time complexity: O(n)
# Space complexity: O(1) - For the set, max of 26 characters.
#
# Runtime: 70 ms, faster than 98.50%
# Memory Usage: 14.1 MB, less than 59.01%
class ListComprehension:
    def firstUniqChar(self, s: str) -> int:
        return min(
            [s.index(char) for char in set(s) if s.count(char) == 1] or [-1]
        )


def test():
    executors = [Solution]
    tests = [
        ["leetcode", 0],
        ["loveleetcode", 2],
        ["aabb", -1],
    ]
    for executor in executors:
        start = timeit.default_timer()
        for _ in range(1):
            for n, t in enumerate(tests):
                sol = executor()
                result = sol.firstUniqChar(t[0])
                exp = t[1]
                assert result == exp, (
                    f"\033[93m» {result} <> {exp}\033[91m for "
                    + f"test {n} using \033[1m{executor.__name__}"
                )
        stop = timeit.default_timer()
        used = str(round(stop - start, 5))
        cols = "{0:20}{1:10}{2:10}"
        res = cols.format(executor.__name__, used, "seconds")
        print(f"\033[92m» {res}\033[0m")


test()

File: leetcode/first-unique-character-in-a-string.rs
// 387. First Unique Character in a String
// 🟢 Easy
//
// https://leetcode.com/problems/first-unique-character-in-a-string/
//
// Tags: Hash Table - String - Queue - Counting

struct Solution;
impl Solution {
    /// Iterate over the input characters saving the index of characters that we see only one time,
    /// use i32::MAX to indicate that we have not seen that character yet, -1 to indicate that we
    /// have seen that character more than one time, and the index if we have seen it only once.
    ///
    /// Time complexity: O(n) - We iterate over the input characters, then over the seen array.
    /// Space complexity: O(1) - We use an array of fixed size 26 of extra memory.
    ///
    /// Runtime 0 ms Beats 100%
    /// Memory 2.28 MB Beats 42.65%
    pub fn first_uniq_char(s: String) -> i32 {
        let mut seen = [i32::MAX; 26];
        let mut cp;
        for (i, ch) in (s.bytes().enumerate()).rev() {
            cp = (ch - b'a') as usize;
            seen[cp] = if seen[cp] == i32::MAX { i as i32 } else { -1 };
        }
        match seen.iter().filter(|&x| *x >= 0 && *x < i32::MAX).min() {
            Some(res) => *res,
            None => -1,
        }
    }
}

// Tests.
fn main() {
    let tests = [("leetcode", 0), ("loveleetcode", 2), ("aabb", -1)];
    println!("\n\x1b[92m» Running {} tests...\x1b[0m", tests.len());
    let mut success = 0;
    for (i, t) in tests.iter().enumerate() {
        let res = Solution::first_uniq_char(t.0.to_string());
        if res == t.1 {
            success += 1;
            println!("\x1b[92m✔\x1b[95m Test {} passed!\x1b[0m", i);
        } else {
            println!(
                "\x1b[31mx\x1b[95m Test {} failed expected: {:?} but got {}!!\x1b[0m",
                i, t.1, res
            );
        }
    }
    println!();
    if success == tests.len() {
        println!("\x1b[30;42m✔ All tests passed!\x1b[0m")
    } else if success == 0 {
        println!("\x1b[31mx \x1b[41;37mAll tests failed!\x1b[0m")
    } else {
        println!(
            "\x1b[31mx\x1b[95m {} tests failed!\x1b[0m",
            tests.len() - success
        )
    }
}
